File: t20-data/src/__tests__/point-buy.test.ts
import { describe, expect, it } from 'vitest'
import {
  POINT_BUY_BUDGET,
  pointBuyCost,
  pointBuySpent,
  pointBuyWarnings,
} from '../point-buy'

const spread = (over: Partial<Record<string, number>>) => ({
  strength: 0,
  dexterity: 0,
  constitution: 0,
  intelligence: 0,
  wisdom: 0,
  charisma: 0,
  ...over,
})

describe('pointBuyCost — Tabela 1-1 (p17)', () => {
  it('custos pinados do livro', () => {
    expect(pointBuyCost(-1)).toBe(-1)
    expect(pointBuyCost(0)).toBe(0)
    expect(pointBuyCost(1)).toBe(1)
    expect(pointBuyCost(2)).toBe(2)
    expect(pointBuyCost(3)).toBe(4)
    expect(pointBuyCost(4)).toBe(7)
  })

  it('fora do intervalo lança com o valor ofensor', () => {
    expect(() => pointBuyCost(5)).toThrow(/got 5/)
    expect(() => pointBuyCost(-2)).toThrow(/got -2/)
  })
})

describe('pointBuySpent + pointBuyWarnings', () => {
  it('spread do exemplo: 4+2+2+1+1+0 = 10 pontos, legal', () => {
    const attrs = spread({
      strength: 3, // 4
      dexterity: 2, // 2
      constitution: 2, // 2
      intelligence: 1, // 1
      wisdom: 1, // 1
    })
    expect(pointBuySpent(attrs)).toBe(POINT_BUY_BUDGET)
    expect(pointBuyWarnings(attrs)).toEqual([])
  })

  it('−1 devolve 1 ponto', () => {
    expect(pointBuySpent(spread({ charisma: -1, strength: 4, dexterity: 2, constitution: 2 }))).toBe(10)
  })

  it('acima do limite avisa', () => {
    const w = pointBuyWarnings(spread({ strength: 4, dexterity: 4 }))
    expect(w.some((x) => x.includes('excedem'))).toBe(true)
  })

  it('dois atributos em −1 avisa (só UM pode, p17)', () => {
    const w = pointBuyWarnings(spread({ charisma: -1, wisdom: -1 }))
    expect(w.some((x) => x.includes('UM atributo'))).toBe(true)
  })

  it('valor fora do intervalo avisa sem lançar', () => {
    const w = pointBuyWarnings(spread({ strength: 5 }))
    expect(w.some((x) => x.includes('fora do intervalo'))).toBe(true)
  })
})

File: engine-go/engine/pointbuy_test.go
package engine

import (
	"reflect"
	"testing"
)

func spread(over map[string]int) map[string]int {
	m := map[string]int{"strength": 0, "dexterity": 0, "constitution": 0, "intelligence": 0, "wisdom": 0, "charisma": 0}
	for k, v := range over {
		m[k] = v
	}
	return m
}

func TestPointBuySpent(t *testing.T) {
	// Book example: 3(4)+2(2)+2(2)+1(1)+1(1) = 10, legal.
	got, ok := PointBuySpent(spread(map[string]int{"strength": 3, "dexterity": 2, "constitution": 2, "intelligence": 1, "wisdom": 1}))
	if !ok || got != PointBuyBudget {
		t.Fatalf("book spread: got (%d, %v), want (10, true)", got, ok)
	}
	// −1 refunds a point.
	if s, ok := PointBuySpent(spread(map[string]int{"charisma": -1, "strength": 4, "dexterity": 2, "constitution": 2})); !ok || s != 10 {
		t.Fatalf("refund spread: got (%d, %v), want (10, true)", s, ok)
	}
	// Out of range → ok=false.
	if _, ok := PointBuySpent(spread(map[string]int{"strength": 5})); ok {
		t.Fatalf("out-of-range spread: ok=true, want false")
	}
}

func TestPointBuyWarnings(t *testing.T) {
	if w := PointBuyWarnings(spread(map[string]int{"strength": 3, "dexterity": 2, "constitution": 2, "intelligence": 1, "wisdom": 1})); len(w) != 0 {
		t.Fatalf("legal spread warnings: %v, want none", w)
	}
	// Exact pt-BR strings (byte-parity with point-buy.ts — note U+2212 in the −1 message).
	overW := PointBuyWarnings(spread(map[string]int{"strength": 4, "dexterity": 4}))
	if !reflect.DeepEqual(overW, []string{"compra de pontos: 14 pontos gastos excedem o limite de 10"}) {
		t.Fatalf("over-budget warnings: %q", overW)
	}
	reducedW := PointBuyWarnings(spread(map[string]int{"charisma": -1, "wisdom": -1}))
	if !reflect.DeepEqual(reducedW, []string{"compra de pontos: apenas UM atributo pode ser reduzido a −1 (p17), há 2"}) {
		t.Fatalf("two-reduced warnings: %q", reducedW)
	}
	rangeW := PointBuyWarnings(spread(map[string]int{"strength": 5}))
	if !reflect.DeepEqual(rangeW, []string{"compra de pontos: strength = 5 fora do intervalo [-1, 4]"}) {
		t.Fatalf("out-of-range warnings: %q", rangeW)
	}
}
